track.go

```diff
@@ -1,76 +1,186 @@
+	return t.id
+}
+
+// PayloadType gets the PayloadType of the track
+func (t *Track) PayloadType() uint8 {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.payloadType
+}
+
+// Kind gets the Kind of the track
+func (t *Track) Kind() RTPCodecType {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.kind
+}
+
+// Label gets the Label of the track
+func (t *Track) Label() string {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.label
+}
+
+// SSRC gets the SSRC of the track
+func (t *Track) SSRC() uint32 {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.ssrc
+}
+
+// Codec gets the Codec of the track
+func (t *Track) Codec() *RTPCodec {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.codec
+}
+
+// Read reads data from the track. If this is a local track this will error
+func (t *Track) Read(b []byte) (n int, err error) {
+	t.mu.RLock()
+	if len(t.senders) != 0 {
+		t.mu.RUnlock()
+		return 0, fmt.Errorf("this is a local track and must not be read from")
 	}
+	r := t.receiver
+	t.mu.RUnlock()
 
-	return &Track{
-		isRawRTP: true,
-
-		ID:          id,
-		PayloadType: payloadType,
-		Kind:        codec.Type,
-		Label:       label,
-		SSRC:        ssrc,
-		Codec:       codec,
-	}, nil
+	return r.readRTP(b)
 }
 
-// NewSampleTrack initializes a new *Track configured to accept media.Sample
-func NewSampleTrack(payloadType uint8, id, label string, codec *RTPCodec) (*Track, error) {
-	if codec == nil {
-		return nil, errors.New("codec supplied to NewSampleTrack() must not be nil")
+// ReadRTP is a convenience method that wraps Read and unmarshals for you
+func (t *Track) ReadRTP(b []byte) (*rtp.Packet, error) {
+	i, err := t.Read(b)
+	if err != nil {
+		return nil, err
 	}
 
-	buf := make([]byte, 4)
-	if _, err := rand.Read(buf); err != nil {
-		return nil, errors.New("failed to generate random value")
+	r := &rtp.Packet{}
+	if err := r.Unmarshal(b[:i]); err != nil {
+		return nil, err
 	}
+	return r, nil
+}
+
+// Write writes data to the track. If this is a remote track this will error
+func (t *Track) Write(b []byte) (n int, err error) {
+	t.mu.RLock()
+	if t.receiver != nil {
+		t.mu.RUnlock()
+		return 0, fmt.Errorf("this is a remote track and must not be written to")
+	}
+	senders := t.senders
+	t.mu.RUnlock()
+
+	for _, s := range senders {
+		if _, err := s.sendRTP(append([]byte{}, b...)); err != nil {
+			return 0, err
+		}
+	}
+
+	return len(b), nil
+}
+
+// WriteSample packetizes and writes to the track
+func (t *Track) WriteSample(s media.Sample) error {
+	packets := t.packetizer.Packetize(s.Data, s.Samples)
+	for _, p := range packets {
+		buf, err := p.Marshal()
+		if err != nil {
+			return err
+		}
+		if _, err := t.Write(append([]byte{}, buf...)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// WriteRTP writes RTP packets to the track
+func (t *Track) WriteRTP(p *rtp.Packet) error {
+	buf, err := p.Marshal()
+	if err != nil {
+		return err
+	}
+	if _, err := t.Write(append([]byte{}, buf...)); err != nil {
```

Comment:
Same comment for this entire file. `Write` should never modify the given buffer.

In fact, this append is completely unnecessary and a good example of why this convention exists.

Reply:
Fixed, SRTP actually allocates for us I forgot https://github.com/pions/webrtc/pull/437/commits/e112cd42c71a9afe8fdc34ec3fc83b4e292270e4